### PR TITLE
fix(dfloat): parse counts significant digits, not digit characters

### LINKED ISSUE #1485: dfloat: parse counts a fraction's leading zeros as significant digits
## Summary

`dfloat::assign()` / `parse()` treat the leading zeros of a fraction as significant digits. A number below 1 with leading zeros loses its trailing digits, and past a point its exponent is wrong too. Measured on main (BID):

| type | input | result | correct |
|---|---|---|---|
| decimal32 | `"0.0001234567"` | `1.23e-8` | `1.234567e-4` |
| decimal32 | `"0.000001234567"` | `1e-12` | `1.234567e-6` |
| decimal64 | `"0.00000000001234567890123456"` | `1.2345e-22` | `1.234567890123456e-11` |
| `dfloat<4,6>` | `"-0.09982"` | `-9.9e-4` | `-0.09982` |
| `dfloat<4,6>` | `"0.00004155"` | `0` | `4.155e-5` |

The same values written with an exponent parse correctly: `"-9982e-5"` gives `-0.09982` and `"4155e-8"` gives `4.155e-5`.

The results suggest the digit count includes the zeros between the decimal point and the first nonzero digit, so the significand is truncated to `ndigits` counted from the decimal point. The exponent is then off by the number of leading zeros. The conversion tests (`string_parse.cpp`) only use inputs such as `"0.5"` and `"0.0625"`, which have at most one leading zero.

## Proposed fix and test

Skip leading zeros (before and after the decimal point) when counting significant digits, and fold them into the exponent only. Regression: round-trip `1.234567 * 10^-k` for k = 0..20 written in positional form, against the exponent form, for decimal32, decimal64 and a narrow width.

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary

Fixes #1485. `dfloat::assign()`, which `parse()` and `operator>>` use, kept the first `ndigits` digit **characters** as the significand, so leading zeros used up the precision and digits past the precision in the fraction still moved the exponent. The two cases in the issue turned out to be part of a wider defect:

| input | main | this PR |
|---|---|---|
| decimal32 `"0.0001234567"` | `1.23e-8` | `1.234567e-4` |
| decimal32 `"0000000123"` | `0` | `123` |
| decimal32 `"1.23456789"` | `0.01234567` | `1.234567` |
| decimal32 `"12345678.9"` | `1234567` | `1.234567e7` |
| decimal64 `"3.14159265358979323846"` | `3.141592653589793e-5` | `3.141592653589793` |
| decimal32 `"1e99999999999"` | signed overflow (UB) | `inf` |

## Fix

The digit loop keeps value = `sig * 10^(decimal_exponent - frac_digits)`, where `frac_digits` counts only the fraction digits actually appended to `sig`.
- Leading zeros take no precision; after the point they only set the scale.
- Digits past the precision are dropped (truncated, as dfloat rounds elsewhere, #1487). A dropped integer digit scales the value up; a dropped fraction digit changes nothing.
- The exponent clamps at exactly 10^8, beyond every dfloat's range even after a long mantissa has offset it, and far from int overflow. (The first version's guard stopped growing somewhere between 10^6 and 9999999, depending on the digits; CodeRabbit caught it, and the test now covers a 10^7-digit mantissa against an 8-digit exponent.)

## Tests

New `dfloat_parse_digits` (`static/float/dfloat/conversion/parse_digits.cpp`):
- **Six textual forms per value.** Random digit strings of 1 to `3*ndigits` digits are each written as: plain positional, with leading integer zeros, with trailing fraction zeros, with the point shifted against an explicit exponent, scientific, and integer significand with an exponent. Every form must parse to the value truncated to `ndigits`, computed independently on the digit strings; `parse()` must accept it and agree with `assign()`. This runs for `dfloat<4>`, decimal32 and decimal64 in BID and DPD, and decimal128 and `dfloat<10>` at higher levels.
- **Positional round trip** of `1.234567e-k` for k = 0 to 20, as the issue proposed.
- **The reported cases**, plus 11-digit exponents.

On main it fails every configuration (1,910 of 3,000 decimal32 forms). With this PR it passes every level on gcc 13.3 and clang with 0 warnings, and the probes run clean under UBSan. All 15 dfloat tests pass, including the existing `string_parse`.

**Nothing that parsed correctly changes.** Over 78,000 parsed forms the fix is right on every one. main was right on 28,806, and every one of those keeps identical stored bits. The dfloat differential against main is byte-identical, because its inputs are integer significands with an exponent, which already parsed correctly.

## Related

It merges cleanly with #1488 (dfloat arithmetic) and #1486 (DPD trailing digits), which edit other parts of `dfloat_impl.hpp`.

BEHAVIOUR-CHANGE: dfloat parsing (assign, parse, operator>>) now reads leading zeros and digits beyond the precision correctly (decimal32 "0.0001234567" was 1.23e-8, "1.23456789" was 0.01234567, "0000000123" was 0), and saturates exponents too long for an int

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01NazYbxaii2XXVzQeQKMDTs

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved decimal arithmetic accuracy for alignment, addition, subtraction, division, and precision reduction.
  - Improved decimal parsing for inputs with leading zeros, fractional values, shifted exponents, and excess precision.
  - Corrected exponent handling when digits are truncated while preserving significant-digit truncation behavior.
  - Prevented overflow-related issues when parsing very large explicit exponents.
  - Addressed edge cases affecting positive, negative, BID, and DPD decimal formats.

- **Tests**
  - Added comprehensive regression coverage for positional, scientific, padded, fractional, and extreme-exponent inputs.

### CODE REVIEW

File: static/float/dfloat/conversion/parse_digits.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+	ReportTestSuiteResults(test_suite, nrOfFailedTestCases);
+	return EXIT_SUCCESS;   // ignore failures
+#else
+
+#if REGRESSION_LEVEL_1
+	nrOfFailedTestCases += ReportTestResult(VerifyReportedCases(reportTestCases), "reported cases", test_tag);
+	nrOfFailedTestCases += ReportTestResult(VerifyNegativePowers<7, 6>(reportTestCases), "decimal32 1.234567e-k positional", test_tag);
+	nrOfFailedTestCases += ReportTestResult(VerifyNegativePowers<16, 8>(reportTestCases), "decimal64 1.234567e-k positional", test_tag);
+	nrOfFailedTestCases += ReportTestResult(VerifyParseForms<4, 6, DecimalEncoding::BID>(500, reportTestCases), "dfloat<4,6,BID> forms", test_tag);
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply static/float/dfloat/conversion/parse_digits.cpp]: Agreed. #1485 asked for a narrow width as well, and the random `dfloat<4>` forms do not guarantee every k. Added `VerifyNegativePowers<4, 6>` in 51c4e018. In `dfloat<4>`, `1.234567` has more digits than the precision, so the round trip also exercises the dropped-digit path. On main all 21 cases fail (`"0.1234567"` parses as `123e-7`), and all pass with the fix on gcc and clang.

File: include/sw/universal/number/dfloat/dfloat_impl.hpp [hunk 2]
```diff
@@ -726,9 +737,11 @@ class dfloat {
 			if (pos < txt.size() && txt[pos] == '-') { exp_neg = true; ++pos; }
 			else if (pos < txt.size() && txt[pos] == '+') { ++pos; }
 
+			// saturate: any exponent this large overflows or underflows every dfloat,
+			// and an unbounded int accumulation would be undefined behaviour
 			int exp_val = 0;
 			while (pos < txt.size() && txt[pos] >= '0' && txt[pos] <= '9') {
-				exp_val = exp_val * 10 + (txt[pos] - '0');
+				if (exp_val < 1000000) exp_val = exp_val * 10 + (txt[pos] - '0');
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/number/dfloat/dfloat_impl.hpp]: Agreed, fixed in 683bf7a2. The guard stopped growing anywhere between 10^6 and 9999999, depending on the digits. The exponent now clamps at exactly the cap with your `exp_val > (cap - digit) / 10` form.

I raised the cap from 10^6 to 10^8: a long mantissa can offset the exponent, and 10^8 stays exact for mantissas of up to ~10^8 digits while remaining far from int overflow. The test now covers the case the old guard got wrong: 10^7 leading fraction zeros against `e+10000001`, which read as 0 instead of 1. UBSan is silent on `1e2147483647` and `1e-2147483648`.

File: static/float/dfloat/conversion/parse_digits.cpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+	ReportTestSuiteResults(test_suite, nrOfFailedTestCases);
+	return EXIT_SUCCESS;   // ignore failures
+#else
+
+#if REGRESSION_LEVEL_1
+	nrOfFailedTestCases += ReportTestResult(VerifyReportedCases(reportTestCases), "reported cases", test_tag);
+	nrOfFailedTestCases += ReportTestResult(VerifyNegativePowers<7, 6>(reportTestCases), "decimal32 1.234567e-k positional", test_tag);
+	nrOfFailedTestCases += ReportTestResult(VerifyNegativePowers<16, 8>(reportTestCases), "decimal64 1.234567e-k positional", test_tag);
+	// a narrow width, where 1.234567 has more digits than the precision
+	nrOfFailedTestCases += ReportTestResult(VerifyNegativePowers<4, 6>(reportTestCases), "dfloat<4,6> 1.234567e-k positional", test_tag);
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply static/float/dfloat/conversion/parse_digits.cpp]: Fixed in 683bf7a2. The file had 19 lines past 120 columns. The `forms[]` initializer now puts each comment on its own line, and the rest went through `clang-format --lines`, which wraps them in the repository's continuation style. A whole-file clang-format would also have reflowed 200+ lines that are within the limit. The file now has no line past 120 columns.
